src/iterator.rs

```diff
@@ -394,317 +649,32 @@ impl Signals {
-            poll.poll(&mut events, Some(Duration::from_secs(10)))
-                .unwrap();
-            let event = events.iter().next().unwrap();
-            assert!(event.readiness().is_readable());
-            assert_eq!(token, event.token());
-            let sig = signals.pending().next().unwrap();
-            assert_eq!(crate::SIGUSR1, sig);
-        }
-    }
-}
-
-#[cfg(any(test, feature = "mio-0_7-support"))]
-mod mio_0_7_support {
-    use std::io::Error;
-    use std::os::unix::io::AsRawFd;
-
-    use mio_0_7::event::Source;
-    use mio_0_7::unix::SourceFd;
-    use mio_0_7::{Interest, Registry, Token};
-
-    use super::Signals;
-
-    impl Source for Signals {
-        fn register(
-            &mut self,
-            registry: &Registry,
-            token: Token,
-            interest: Interest,
-        ) -> Result<(), Error> {
-            SourceFd(&self.waker.read.as_raw_fd()).register(registry, token, interest)
-        }
-
-        fn reregister(
-            &mut self,
-            registry: &Registry,
-            token: Token,
-            interest: Interest,
-        ) -> Result<(), Error> {
-            SourceFd(&self.waker.read.as_raw_fd()).reregister(registry, token, interest)
-        }
-
-        fn deregister(&mut self, registry: &Registry) -> Result<(), Error> {
-            SourceFd(&self.waker.read.as_raw_fd()).deregister(registry)
-        }
-    }
-
-    #[cfg(test)]
-    mod tests {
-        use std::time::Duration;
-
-        use mio_0_7::{Events, Poll};
-
-        use super::*;
-
-        #[test]
-        fn mio_wakeup() {
-            let mut signals = Signals::new(&[crate::SIGUSR1]).unwrap();
-            let mut poll = Poll::new().unwrap();
-            let token = Token(0);
-            poll.registry()
-                .register(&mut signals, token, Interest::READABLE)
-                .unwrap();
-
-            let mut events = Events::with_capacity(10);
-            unsafe { libc::raise(crate::SIGUSR1) };
-            poll.poll(&mut events, Some(Duration::from_secs(10)))
-                .unwrap();
-            let event = events.iter().next().unwrap();
-
-            assert!(event.is_readable());
-            assert_eq!(token, event.token());
-            let sig = signals.pending().next().unwrap();
-            assert_eq!(crate::SIGUSR1, sig);
+        match self.poll_signal(&mut Signals::has_signals) {
+            PollResult::Signal(result) => Some(result),
+            PollResult::Closed => None,
+            PollResult::Pending => unreachable!(
+                "Because of the blocking has_signals method the \
+                next_signal method never returns Poll::Pending but blocks until a signal arrived"
+            ),
+            // Users can't manipulate the internal file descriptors and the way we use them
+            // shouldn't produce any errors. So it is OK to panic.
+            PollResult::Err(error) => panic!("Unexpected error: {}", error),
```

Comment:
I feel like I've seen a very similar match somewhere up there. How hard would it be to reuse it, would it add or hinder readability and maintainability?

Reply:
There are some similarities between the matches and two of their right hand sides are identical. But the matched values are of different types. So the only possible reuse I see is moving the two identical parts into two functions and use those in the match expression. This would de-duplicate the string literals in the unreachable! and panic! macros.